lib/src/subtree.h

```diff
@@ -39,21 +40,56 @@ typedef struct {
 //
 // This representation is used for small leaf nodes that are not
 // errors, and were not created by an external scanner.
+// The idea behind the layout of this struct is that the `is_inline`
+// bit will fall exactly into the same location as the least significant
+// bit of the pointer in `Subtree` or `MutableSubtree`, respectively.
+// Because of alignment, for any valid pointer this will be 0, giving
+// us the opportunity to make use of this bit to signify whether to use
+// the pointer or the inline struct.
 typedef struct {
-  bool is_inline : 1;
-  bool visible : 1;
-  bool named : 1;
-  bool extra : 1;
-  bool has_changes : 1;
-  bool is_missing : 1;
+#define SUBTREE_3_BYTES \
+  uint8_t symbol; \
+  uint8_t padding_bytes; \
+  uint8_t size_bytes; \
+
+#define SUBTREE_6_BITS \
+  bool visible : 1; \
+  bool named : 1; \
+  bool extra : 1; \
+  bool has_changes : 1; \
+  bool is_missing : 1; \
   bool is_keyword : 1;
-  uint8_t symbol;
-  uint8_t padding_bytes;
-  uint8_t size_bytes;
-  uint8_t padding_columns;
-  uint8_t padding_rows : 4;
-  uint8_t lookahead_bytes : 4;
+
+#define SUBTREE_4_BYTES \
+  uint8_t padding_columns; \
+  uint8_t padding_rows : 4; \
+  uint8_t lookahead_bytes : 4; \
   uint16_t parse_state;
```

Comment:
Some minor naming suggestions:

1. If you switched the `uint16_t parse_state;` with `uint8_t padding_bytes; uint8_t size_bytes;`, then all of the "size"-related fields would be together in `SUBTREE_4_BYTES`, and you could rename that macro to `SUBTREE_SIZE`, which indicates more about its meaning.
2. Then, `SUBTREE_3_BYTES` would only contain two fields, and you could just inline them, instead of having that macro at all.
2. I think `SUBTREE_6_BITS` could be renamed to just `SUBTREE_BITS`.

Reply:
The idea behind the names was that each indicates exactly how much space it takes so the layout is directly visible. But your suggestion works too, so I have updated it accordingly.